packages/auth/src/core/auth/auth_impl.ts

```diff
@@ -324,14 +325,26 @@ export class AuthImpl implements AuthInternal, _FirebaseService {
         AuthErrorCode.TENANT_ID_MISMATCH
       );
     }
+    await this._runBeforeStateCallbacks(user);
 
     return this.queue(async () => {
       await this.directlySetCurrentUser(user as UserInternal | null);
       this.notifyAuthListeners();
     });
   }
 
+  async _runBeforeStateCallbacks(user: User | null): Promise<void> {
+    try {
+      for (const beforeStateCallback of this.beforeStateQueue) {
+        await beforeStateCallback(user);
+      }
+    } catch (e) {
+      throw this._errorFactory.create(AuthErrorCode.LOGIN_BLOCKED, { message: e.message });
+    }
+  }
+
   async signOut(): Promise<void> {
+    await this._runBeforeStateCallbacks(null);
```

Comment:
This will run twice since it's also in _updateCurrentUser

Reply:
It seems like it needs to run before `_setRedirectUser()` so it can block it, so I added an optional arg to `_updateCurrentUser()` to skip running the callbacks inside it.